src/quantifier.ml

```diff
@@ -1030,703 +1423,2007 @@ let select_implicant srk interp ?(env=Env.empty) phi =
     Some (specialize_floor_cube srk interp atoms)
   | None -> None
 
+(** The type of modules implementing Strategy Improvement *)
+module type StrategyImprovement = sig
+  (** Satisfiability via strategy improvement *)
+  val simsat : 'a context -> 'a formula -> [ `Sat | `Unsat | `Unknown ]
 
-(* Counter-strategy synthesis *)
-module CSS = struct
+  (** Satisfiability via strategy improvement, forwards version *)
+  val simsat_forward : 'a context -> 'a formula -> [ `Sat | `Unsat | `Unknown ]
 
-  type 'a ctx =
-    { formula : 'a formula;
-      not_formula : 'a formula; (* Negated formula *)
-      mutable skeleton : Skeleton.t; (* skeleton for formula *)
+  (** Alternating quantifier optimization *)
+  val maximize : 'a context -> 'a formula -> 'a arith_term -> [ `Bounded of QQ.t
+                                                              | `Infinity
+                                                              | `MinusInfinity
+                                                              | `Unknown ]
+  (** Alternating quantifier satisfiability *)
+  val easy_sat : 'a context -> 'a formula -> [ `Sat | `Unsat | `Unknown ]
 
-      (* solver for the *negation* of the winning formula for skeleton (unsat
-         iff there is a winning SAT strategy for formula which conforms to
-         skeleton) *)
-      solver : 'a Smt.Solver.t;
-      srk : 'a context;
-    }
+  type 'a skeleton
+  type 'a move
+  type 'a strategy
+  type 'a normalized_formula
 
-  let reset ctx =
-    Smt.Solver.reset ctx.solver;
-    ctx.skeleton <- Skeleton.SEmpty
+  val pp_skeleton : 'a context -> Format.formatter -> 'a skeleton -> unit
+  val show_skeleton : 'a context -> 'a skeleton -> string
 
-  let add_path ctx path =
-    let srk = ctx.srk in
-    try
-      ctx.skeleton <- Skeleton.add_path srk path ctx.skeleton;
-      let win =
-        Skeleton.path_winning_formula srk path ctx.skeleton ctx.formula
+  val winning_skeleton : 'a context -> 'a normalized_formula ->
+    [ `Sat of 'a skeleton
+    | `Unsat of 'a skeleton
+    | `Unknown ]
+
+  val pp_strategy : 'a context -> Format.formatter -> 'a strategy -> unit
+  val show_strategy : 'a context -> 'a strategy -> string
+
+  (** Compute a winning SAT/UNSAT strategy for a normalized formula *)
+  val winning_strategy : 'a context -> 'a normalized_formula ->
+      [ `Sat of 'a strategy
+      | `Unsat of 'a strategy
+      | `Unknown ]
+
+  (** Check that a SAT strategy is in fact a winning Strategy. (To verify
+      an UNSAT stragey provide the negation of the normalized function) *)
+  val check_strategy : 'a context -> 'a normalized_formula -> 'a strategy -> bool
+
+  (** take a formula and return a normalized formula *)
+  val normalize : 'a context -> 'a formula -> 'a normalized_formula
+
+  val pp_normal : 'a context -> Format.formatter -> 'a normalized_formula -> unit
+end
+
+type 'a cg_formula = ([`Forall | `Exists] * symbol) list * 'a formula
+module CoarseGrainStrategyImprovement = struct
```

Comment:
Is there any reason to keep this as a separate implementation?  (E.g., if I convert to prenex and then compute a fg_formula that's just a sequence of quantifiers followed by an atom, and then compute a fine-grained strategy, is that the same as computing a coarse-grained strategy?)  On the other hand, I recall that the story for *non-forwards* fine-grained strategy improvement is complicated and the algorithm is dominated by the forwards variant.  So maybe it would make to have (non-forwards) coarse grained + (forwards) fine grained.  Hopefully we can also get some code re-use too.  E.g., if you use polymorphic variants to define skeletons, then coarse-grained skeletons are a subtype of fine-grained skeletons, so we can re-use fine-grained code for coarse-grained.

Reply:
It's been a while since I read through the coarse-grained implementation. The fine-grained implementation can certainly produce coarse-grained strategies in the construction you gave. At an algorithmic level the fine-grained implementation should behave the same as the coarse-grained implementation for formulas in prenex-normal form. But I'm unsure if the coarse-grained implementation has any optimizations that require the formula be in prenex-normal form (e.g. cannot be used for the fine-grained implementation).

I can review the coarse-grained implementation and then suggest either keeping just the fine-grained or the two variants you suggested.